engine/engine_test.go

```diff
@@ -349,6 +352,206 @@ func TestEngine_NotifyPutThenNotifyRemove(t *testing.T) {
+	_, err = ds.Get(ctx, datastore.NewKey("map/keyCid/"+provider2ID.String()+"/fish"))
+	require.Error(t, err, datastore.ErrNotFound)
+	_, err = ds.Get(ctx, datastore.NewKey("map/keyMD/"+provider2ID.String()+"/fish"))
+	require.Error(t, err, datastore.ErrNotFound)
+
+	// there should be only 2 records left in the cidProvAndKey index for the "fish" fontext id
+	_, c, err = cid.CidFromBytes(cidBytes)
+	require.NoError(t, err)
+	pAndCBytes, err = ds.Get(ctx, datastore.NewKey("map/cidProvAndKey/"+c.String()))
+	require.NoError(t, err)
+
+	pAndC = &engine.ProviderAndContext{}
+	err = json.Unmarshal(pAndCBytes, &pAndC)
+	require.NoError(t, err)
+
+	require.Equal(t, commonContextID, pAndC.ContextID)
+	require.Equal(t, provider3Bytes, pAndC.Provider)
+
+	pAndC = pAndC.Next
+	require.Equal(t, commonContextID, pAndC.ContextID)
+	require.Equal(t, provider1Bytes, pAndC.Provider)
+	require.Nil(t, pAndC.Next)
+
+	// REMOVE ALL - removing all records for provider 3
+	_, err = subject.NotifyRemove(ctx, provider3ID, nil)
+	require.NoError(t, err)
+	_, err = ds.Get(ctx, datastore.NewKey("map/keyCid/"+provider3ID.String()+"/fish"))
+	require.Error(t, err, datastore.ErrNotFound)
+	_, err = ds.Get(ctx, datastore.NewKey("map/keyCid/"+provider3ID.String()+"/bird"))
+	require.Error(t, err, datastore.ErrNotFound)
+	_, err = ds.Get(ctx, datastore.NewKey("map/keyMD/"+provider3ID.String()+"/fish"))
+	require.Error(t, err, datastore.ErrNotFound)
+	_, err = ds.Get(ctx, datastore.NewKey("map/keyMD/"+provider3ID.String()+"/bird"))
+	require.Error(t, err, datastore.ErrNotFound)
+
+	// there should be only 1 record in the cidProvAndKey index left for the "fish" context
+	_, c, err = cid.CidFromBytes(cidBytes)
+	require.NoError(t, err)
+	pAndCBytes, err = ds.Get(ctx, datastore.NewKey("map/cidProvAndKey/"+c.String()))
+	require.NoError(t, err)
+
+	pAndC = &engine.ProviderAndContext{}
+	err = json.Unmarshal(pAndCBytes, &pAndC)
+	require.NoError(t, err)
+
+	require.Equal(t, commonContextID, pAndC.ContextID)
+	require.Equal(t, provider1Bytes, pAndC.Provider)
+	require.Nil(t, pAndC.Next)
+
+	// there should be no records in the cidProvAndKey index left for the "bird" context
+	_, c, err = cid.CidFromBytes(cidBytes4)
+	require.NoError(t, err)
+	_, err = ds.Get(ctx, datastore.NewKey("map/cidProvAndKey/"+c.String()))
+	require.Error(t, err, datastore.ErrNotFound)
+
+	// REMOVE ONE - removing a record for the default provider
+	_, err = subject.NotifyRemove(ctx, "", commonContextID)
+	require.NoError(t, err)
+	_, err = ds.Get(ctx, datastore.NewKey("map/keyCid/fish"))
+	require.Error(t, err, datastore.ErrNotFound)
+	_, err = ds.Get(ctx, datastore.NewKey("map/keyMD/fish"))
+	require.Error(t, err, datastore.ErrNotFound)
+
+	// there should be no records left in the cidProvAndKey index for the "fish" fontext id
+	_, c, err = cid.CidFromBytes(cidBytes)
+	require.NoError(t, err)
+	_, err = ds.Get(ctx, datastore.NewKey("map/cidProvAndKey/"+c.String()))
+	require.Error(t, err, datastore.ErrNotFound)
```

Comment:
any chance we could break up this test in to explicit test cases?

Reply:
broke up into 2 parts, good point

Reply:
done